interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,966 @@
+// saves the database to disk.
+func (rdb *RuleDB) RemoveRule(user uint32, id prompting.IDType) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	rule, err := rdb.ruleWithIDForUser(user, id)
+	if err != nil {
+		return nil, err
+	}
+	err = rdb.removeRuleFromTree(rule)
+	// If error occurs, rule was still fully removed from tree
+	rdb.removeRuleWithID(id)
+	rdb.save()
+	data := map[string]string{"removed": "removed"}
+	rdb.notifyRule(user, id, data)
+	return rule, err
+}
+
+// RemoveRulesForSnap removes all rules pertaining to the given snap for the
+// user with the given user ID.
+func (rdb *RuleDB) RemoveRulesForSnap(user uint32, snap string) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap
+	}
+	rules := rdb.rulesInternal(ruleFilter)
+	rdb.removeRulesInternal(user, rules)
+	return rules
+}
+
+// removeRulesInternal removes all of the given rules from the rule DB and
+// records a notice for each one.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) removeRulesInternal(user uint32, rules []*Rule) {
+	for _, rule := range rules {
+		rdb.removeRuleFromTree(rule)
+		// If error occurs, rule was still fully removed from tree
+		rdb.removeRuleWithID(rule.ID)
+		data := map[string]string{"removed": "removed"}
+		rdb.notifyRule(user, rule.ID, data)
+	}
+	rdb.save()
+}
+
+// RemoveRulesForSnapInterface removes all rules pertaining to the given snap
+// and interface for the user with the given user ID.
+func (rdb *RuleDB) RemoveRulesForSnapInterface(user uint32, snap string, iface string) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap && rule.Interface == iface
+	}
+	rules := rdb.rulesInternal(ruleFilter)
+	rdb.removeRulesInternal(user, rules)
+	return rules
+}
+
+// PatchRule modifies the rule with the given ID by updating the rule's fields
+// corresponding to any of the given parameters which are set/non-empty.
+//
+// Any of the parameters which are equal to the default/unset value for their
+// types are left unchanged from the existing rule. Even if the given new rule
+// contents exactly match the existing rule contents, the timestamp of the rule
+// is updated to the current time. If there is any error while modifying the
+// rule, the rule is rolled back to its previous unmodified state, leaving the
+// database unchanged. If the database is changed, it is saved to disk.
+func (rdb *RuleDB) PatchRule(user uint32, id prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	changeOccurred := false
+	defer func() {
+		switch {
+		case changeOccurred:
+			rdb.notifyRule(user, id, nil)
+			fallthrough
+		case needToSave:
+			rdb.save()
```

Comment:
probably we have the same question here where we might save on error but unclear why?